ts/test/session/sending/MessageQueue_test.ts

```diff
@@ -0,0 +1,202 @@
+} from '../../test-utils/testUtils';
+import { OpenGroupMessage } from '../../../session/messages/outgoing';
+import { PubKey, RawMessage } from '../../../session/types';
+import { UserUtil } from '../../../util';
+import { MessageSender } from '../../../session/sending';
+import { toRawMessage } from '../../../session/utils/Messages';
+import { SessionProtocol } from '../../../session/protocols';
+import { PendingMessageCache } from '../../../session/sending/PendingMessageCache';
+
+// Equivalent to Data.StorageItem
+interface StorageItem {
+  id: string;
+  value: any;
+}
+
+describe('MessageQueue', () => {
+  // Initialize new stubbed cache
+  let data: StorageItem;
+  const sandbox = sinon.createSandbox();
+  const ourNumber = generateFakePubkey().key;
+
+  // Keep track of Session Requests in each test
+  let sessionRequestSent: boolean;
+
+  // Initialize new stubbed cache
+  let messageQueueStub: MessageQueue;
+  // Message Sender Stubs
+  let sendStub: sinon.SinonStub<[RawMessage, (number | undefined)?]>;
+  let sendToOpenGroupStub: sinon.SinonStub<[OpenGroupMessage]>;
+  // Group Utils Stubs
+  let isMediumGroupStub: sinon.SinonStub;
+  let groupMembersStub: sinon.SinonStub;
+  // Session Protocol Stubs
+  let hasSessionStub: sinon.SinonStub;
+
+  beforeEach(async () => {
+    // Stub out methods which touch the database
+    const storageID = 'pendingMessages';
+    data = {
+      id: storageID,
+      value: '[]',
+    };
+
+    // Pending Message Cache Data Stubs
+    TestUtils.stubData('getItemById')
+      .withArgs('pendingMessages')
+      .callsFake(async () => {
+        return data;
+      });
+    TestUtils.stubData('createOrUpdateItem').callsFake((item: StorageItem) => {
+      if (item.id === storageID) {
+        data = item;
+      }
+    });
+
+    // Utils Stubs
+    sandbox.stub(UserUtil, 'getCurrentDevicePubKey').resolves(ourNumber);
+    TestUtils.stubData('getPairedDevicesFor').callsFake(async () => {
+      return generateMemberList(2);
+    });
+    TestUtils.stubWindow('libsignal', {
+      SignalProtocolAddress: sandbox.stub(),
+      SessionCipher: Stubs.SessionCipherStub,
+    } as any);
+
+    // Message Sender Stubs
+    sendStub = sandbox.stub(MessageSender, 'send').resolves();
+    sendToOpenGroupStub = sandbox
+      .stub(MessageSender, 'sendToOpenGroup')
+      .resolves(true);
+
+    // Group Utils Stubs
+    isMediumGroupStub = sandbox
+      .stub(GroupUtils, 'isMediumGroup')
+      .resolves(false);
+    groupMembersStub = sandbox
+      .stub(GroupUtils, 'getGroupMembers' as any)
+      .callsFake(async () => generateMemberList(10));
+
+    // Session Protocol Stubs
+    hasSessionStub = sandbox.stub(SessionProtocol, 'hasSession').resolves(true);
+    sandbox.stub(SessionProtocol, 'sendSessionRequest').resolves();
+    sandbox
+      .stub(SessionProtocol, 'sendSessionRequestIfNeeded')
+      .callsFake(async (pubkey: PubKey) => {
+        pubkey;
+        sessionRequestSent = true;
```

Comment:
you can also do this via the stub itself.
```
const stub = <stub>.called
```
but it's minor so it'll be up to you if you want to change it